README.md
# CrossCorrelation
Pakcage to do the cross-correlation delta - QSO

To get on your comuter:
    clone https://_YOUR_LOGIN_NAME@github.com/londumas/CrossCorrelation.git



CHANGELOG
# 2016 - 01 - 29
	- PYTHON classes :
		* correct bugs relative to pointers. Now do a "copy.deepcopy()" of dictionnaires
		* add a "get_best_fit" method
		* add geting the mean over manu simulations
		* add a empty data method
		* correct the "set_error_on_covar_matrix" method to take into acount multipol and wedge

	- Correlation.cpp
		* remove "1D" in "xi_delta_QSO_lambda" method
# 2016 - 01 - 29
	- myTools.py:
		* add a function to append FITS files
	- Correlation.cpp:
		* add dealing with co-added files
	- GetDelta.cpp:
		* correct bug in putReObsTogether, when testing the coordinate

# 2016 - 02 - 08
	- /Mock_JMLG/ReadFits/src/GetDelta.cpp:
		* correct bug when reading a double instead of a float
	- /Python/Correlation/correlation_3D.py b/Python/Correlation/correlation_3D.py
		* add saving the grid
		* apply distortion matrix to CAMB instead of to data
		* add a plot_slice_2d function to see slice of the 2D correlations

# 2016 - 02 - 09
	* Add calcul of the metal templates and its corresponding PYTHON reading function.

# 2016 - 02 - 11
	- correlation_1D.py
		* Add printing relevant information on init
	- correlation_3D.py
		* Add printing relevant information on init
		* add a "fit_CAMB_2d" method, not perfect yet
		* sort output for metals
		* add a "plot_1d_with_residuals" method
# 2016 - 02 - 11
	- GetDelta.cpp
		* correct error with mean_lambda_rf

# 2016 - 02 - 16
	- Add reading the raw files from Jean-Marc
	- bigger marker and lines in all Python
	- correct a possible error with r > 200 in "annalyse_BAOFIT.py"
	- "Correlation3D" class:
		* add a "read_data_from_BAOFIT_data_file" method to the "Correlation3D" class
		* correct .ini file generation file with metals
	- Correlation.cpp:
		* no more float
		* put long double sometime
		* correct error for the distortion matrix, when reading the lambda_rf value
	- GetDelta.cpp:
		* no more float
		* correct small error when getting mean_lambda_rf

# 2016 - 02 - 17
	- correlation_3D_Q.py:
		* finished and corrected some bugs, still one with multipol.
# 2016 - 02 - 17
	- correlation_3D_Q.py:
		* Correct bog with multipol linked to the fact that with simulations the 'mu'
			coordinate has no meaning.
# 2016 - 02 - 17
	- Correlation.cpp:
		* Remove saving xi_q_q_1D, becuase information is already in xi_q_q_mu
# 2016 - 02 - 17
	- Correlation.cpp:
		* Add conting removed forest
		* Add Nicolas projection for the second set of forests
	- GetDelta.cpp:
		* Add conting removed forest
		* Add calculating mean_lambda_rf when loading the data
		* Add calculating and saving mean_lambda_rf even for bad forests
# 2016 - 02 - 18
	- Correlation.cpp:
		* Correct randomization of QSOs positions in cells for the QSO-QSO correlation
			in Jean6marc's mocks
		* remove useless « randomPositionOfQSOInCell » variable
	- correlation_3D_Q.py:
		* correct calculation of coordinate position for xiMu
# 2016 - 02 - 18
	- GetDelta.cpp:
		* Correct error induced by calculating the < lambda_RF > at each step.
			I calcul it only when saving and updating the delta, at the end.
			Now the alpha and beta converges

# 2016 - 02 - 18
	- GetDelta.cpp:
		* Add 100 bins at the end of "hDeltaVsLambdaObs__" in order to avoid
			the interpolation problem at the edge.
# 2016 - 02 - 18 - (10h08)
	- Correlation.cpp:
		* Complete the xi_QSO_QSO function for data in order to have my one (but crappy)
			auto correlation of QSOs
# 2016 - 02 - 22 - (09h07)
	- everywhere:
		* change the precision to something not arbitrary but to the full precision of a double
	- Cpp/Tools:
		* add a "look_DLA" method to see how the delta of forests with DLA are.
	- Correlation.cpp:
		* correct the way to find "< <delta> >" and "< coef >"
	- GetDelta.cpp:
		* correct the way to deal with DLA. Correct the templat instead of the delta, and fit for it also
		* set to zero very high DLA and set to one very low DLA
# 2016 - 02 - 23 - (09h45)
	- deleted useless files.
	- constants.h:
		* change speed of light dimension
		* change h0 value
	- tools.cpp:
		* add a "get_delta_nicolas" method
	- Mock_JMC:
		* add plate, mjd, fiber
		* only dealing with 647 pixels as data
	- Correlation.h:
		* change size cell for mocks
	- GetDelta.cpp:
		* add pixels low and high for interpolation
		* change value of lambda to lambda-<lambda> in fits
# 2016 - 02 - 23 - (17h47)
	- Correlation.cpp:
		* add option to save some plots in a .root
# 2016 - 02 - 26 - (17h21)
	- Constants.h:
		* add definition of the LYA line
	- Correct value of the lambda_rf_line_LYA in "VoigtProfile" function in order to stay on LYA
	- Mock_JMLG:
		* correct small bug removing forests with no normalisation zone
	- Correlation.cpp
		* add not saving when no pairs in "xi_delta_QSO"
	- GetDelta.cpp
		* big changes! Now the template is like a "delta+1" convergence. Now delta is flat both
			in lamnda_RF and lambda_obs 
# 2016 - 03 - 02 - (10h51)
	- many updates on files to annalyse the results
	- annalyse_BAOFIT_Q.py:
		* add file, in order to annalyse BAOFIT results for QSO-QSO correlations
	- main_Q.py
		* add file to send QSO-QSO correlations
# 2016 - 03 - 04 - (10h28)
	- change format of delta.fits files in order to weight less.
# 2016 - 03 - 08 - (09h15)
	- general:
		* change the structure of the *.fits file to weigth less 
	- Correlation.cpp
		* correct saving Mu and map fore f_f2
		* add options for raw_mocks

# 2016 - 03 - 30 - (10h15)
	- general:
		* no commit for a long time, no big changes.
	- Mock_JMLG/ReadFits/src/GetDelta.cpp:
		* remove "isnan" 

# 2016 - 03 - 31 - (10h16)
	- chain_annalys_delta/Correlation/src/Correlation.cpp:
		* add the metal calculation for LYA and correct the linear templates of CAMB
	- Python/Correlation/CAMB.py:
		* add calculation of monopole, quadupole, hexadecapole.

# 2016 - 05 - 02 - (09h50)
	- /Mock_JMLG/ReadFits/src/GetDelta.cpp:
		* read the files in the order of produced with « ls -tr »
	- /Constants/constants.h
		* change CAMB file to CAMB at z=0, and add a calcul of the growth factor.

Python/QSO/main.py
# -*- coding: utf-8 -*-
#
# created by Hélion du Mas des Bourboux
# < helion.du-mas-des-bourboux@cea.fr >

### My tools
import myTools


import astropy.io.fits as pyfits
import math
import numpy
import os
import decimal ## To set the precision of the double values
import cosmolopy.distance as cosmology
import matplotlib.pyplot as plt

import myTools
from myTools import Get_TProfile
from const_delta import *

def main():
	'''
	'''

	zKEY = 'Z'
	cat = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_ALL_TESTS.fits')[1].data
	cat = cat[ cat['Z']>1.7 ]
	cat = cat[ cat['Z']<7. ]
	cat2 = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR7_DR12_EBOSS.fits')[1].data
	cat2 = cat2[ cat2['Z']>1.7 ]
	cat2 = cat2[ cat2['Z']<7. ]
	cat3 = pyfits.open('/home/gpfs/manip/mnt/bao/hdumasde/Data/LYA/FitsFile_DR12_Guy/DR12_primery/DR12_primery.fits',memmap=True)[1].data[:1000]
	cat3 = cat3[ cat3['Z_VI']>1.7 ]
	cat3 = cat3[ cat3['Z_VI']<7. ]

	print cat.size
	print cat2.size
	print cat3.size
	
	## Map
	plt.plot(cat['RA'],  cat['DEC'], linestyle="", marker="o",label='DR7+DR12: QSO')
	plt.plot(cat3['RA'],  cat3['DEC'], linestyle="", marker="o",label='DR7+DR12: Forest',color='red')
	plt.xlabel(r'$R.A. (\degree)$')
	plt.ylabel(r'$Dec. (\degree)$')
	myTools.deal_with_plot(False,False,True)
	plt.show()
	## Distribution redshift
	plt.hist(cat[zKEY], bins=50,label='DR7+DR12: QSO',histtype='step')
	plt.hist(cat2[zKEY], bins=50,label='DR7+DR12+eBOSS: QSO',histtype='step')
	plt.hist(cat3['Z_VI'], bins=50,label='DR7+DR12: Forest',histtype='step')
	plt.xlabel(r'$z$')
	plt.ylabel(r'$\#$')
	myTools.deal_with_plot(False,False,True)
	plt.show()

	'''
	col_ra              = pyfits.Column(name='RA',  format='D', array=cat['RA'], unit='deg')
	col_de              = pyfits.Column(name='DEC', format='D', array=cat['DEC'], unit='deg')
	col_zz              = pyfits.Column(name='Z',   format='D', array=cat['Z_BEST'] )
	
	tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
	tbhdu.writeto('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_3DHST.fits', clobber=True)
	'''
	return
def getQsoCatalogueDR7():
	'''
		Get the catalogue of qsos for DR7
	'''
	
	redMin = 0.
	redMax = 20.
	
	redShiftKey = "Z"
	
	cat = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/knownquasarstar.060910.fits')[1].data
	redShiftKey = "ZEM"
	
	return numpy.asarray(zip(cat["RA"],cat["DEC"],cat["ZEM"],cat['GMAG']))

	print "  --- DR7 ---"
	print " Size of cat = " + str(len(cat))
	
	cat = cat[ cat[redShiftKey] > redMin ]
	print " Size of cat after cut in Z = " + str(len(cat)) + " qso with z > " + str(redMin)
	
	cat = cat[ cat[redShiftKey] < redMax ]
	print " Size of cat after cut in Z = " + str(len(cat)) + " qso with z < " + str(redMax)
	print " Size of cat after cut in Z = " + str(len(cat[ cat[redShiftKey] >= 2. ])) + " qso with z >= " + str(2.)
	
	file_name = '/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR7.txt'
	file = open(file_name,'w')
	numpy.savetxt(file,zip(cat["RA"],cat["DEC"],cat[redShiftKey]))
	file.close()
	
	print min(cat[redShiftKey])
	print max(cat[redShiftKey])
	
	## Map
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.plot(cat["RA"], cat["DEC"], linestyle="", marker="o")
	plt.xlabel("Right Ascension (degree)")
	plt.ylabel("Declination (degree)")
	plt.title("SDSS DR7")
	plt.show()
	## Distribution redshift
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.hist(cat[redShiftKey], bins=50)
	plt.xlabel("Z")
	plt.ylabel("#")
	plt.title("SDSS DR7")
	plt.show()
	
	return
def getQsoCatalogueDR12():
	'''
		Get the catalogue of qsos
	'''

	redMin = 0.
	redShiftKey = "Z_VI"
	
	cat = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/DR12Q_v2_10.fits')[1].data

	return numpy.asarray(zip(cat["RA"],cat["DEC"],cat["Z_VI"],cat['PSFMAG'][:,1]))
	print "  --- DR12 ---"
	print " Size of cat = " + str(len(cat))
	
	cat = cat[ cat[redShiftKey] > redMin ]
	print " Size of cat after cut in Z = " + str(len(cat)) + " qso with z > " + str(redMin)
	print " Size of cat after cut in Z = " + str(len(cat[ cat[redShiftKey] >= 2. ])) + " qso with z >= " + str(2.)
	
	'''
	file_name = '/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR12.txt'
	file = open(file_name,'w')
	numpy.savetxt(file,zip(cat["RA"],cat["DEC"],cat[redShiftKey]))
	file.close()
	'''
	
	print min(cat[redShiftKey])
	print max(cat[redShiftKey])
	
	## Map
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.plot(cat["RA"], cat["DEC"], linestyle="", marker="o")
	plt.xlabel("Right Ascension (degree)")
	plt.ylabel("Declination (degree)")
	plt.title("BOSS DR12")
	plt.show()
	## Distribution redshift
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.hist(cat[redShiftKey], bins=50)
	plt.xlabel("Z")
	plt.ylabel("#")
	plt.title("BOSS DR12")
	plt.show()
	
	return
def getQsoCatalogueEBOSS():
	'''
		Get the catalogue of qsos
	'''

	path = '/home/gpfs/manip/mnt0607/bao/Spectra/spAll-v5_9_1.fits'
	redShiftKey = 'Z'
	cat = pyfits.open(path, memmap=True )[1].data

	print "  The size of the catalogue is           : ", cat.size
	cat = cat[ cat['CLASS'] == 'QSO' ]
	print '  We keep CLASS == QSO, the size is      : ', cat.size
	cat = cat[ cat['ZWARNING'] == 0 ]
	print '  We keep ZWARNING == 0, the size is      : ', cat.size
	cat = cat[ cat['Z_ERR'] > 0 ]
	print '  We keep Z_ERR > 0, the size is      : ', cat.size

	print  cat[ cat['Z'] > 1.7 ].size


	### BOSS_TARGET1
	selection_BOSS_TARGET1 = [10,11,12,13,14,15,16,17,18,19,40,41,42,43,44]
	bits_BOSS_TARGET1 = 0
	for el in selection_BOSS_TARGET1:
		bits_BOSS_TARGET1 += 2**el
	### ANCILLARY_TARGET1
	selection_ANCILLARY_TARGET1 = [6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,29,30,31,50,51,52,53,54,55,58,59]
	bits_ANCILLARY_TARGET1 = 0
	for el in selection_ANCILLARY_TARGET1:
		bits_ANCILLARY_TARGET1 += 2**el
	### ANCILLARY_TARGET2
	selection_ANCILLARY_TARGET2 = [0,1,2,3,4,5,7,8,9,10,13,14,15,24,25,26,27,31,32,33,53,54,55,56]
	bits_ANCILLARY_TARGET2 = 0
	for el in selection_ANCILLARY_TARGET2:
		bits_ANCILLARY_TARGET2 += 2**el
	### EBOSS_TARGET0
	selection_EBOSS_TARGET0 = [10,11,12,13,14,15,16,17,18,20,22,30,31,33,34,35,40]
	bits_EBOSS_TARGET0 = 0
	for el in selection_EBOSS_TARGET0:
		bits_EBOSS_TARGET0 += 2**el
	### EBOSS_TARGET1
	selection_EBOSS_TARGET1 = [9,10,11,12,13,14,15,16,17,18,30,31]
	bits_EBOSS_TARGET1 = 0
	for el in selection_EBOSS_TARGET1:
		bits_EBOSS_TARGET1 += 2**el
	### EBOSS_TARGET2
	selection_EBOSS_TARGET2 = [0,2,4,20,21,23,24,25,26,27,31]
	bits_EBOSS_TARGET2 = 0
	for el in selection_EBOSS_TARGET2:
		bits_EBOSS_TARGET2 += 2**el

	print cat.size
	select = ((cat['BOSS_TARGET1'] & bits_BOSS_TARGET1 )>0 )            \
		| ((cat['ANCILLARY_TARGET1'] & bits_ANCILLARY_TARGET1 )>0)  \
		| ((cat['ANCILLARY_TARGET2'] & bits_ANCILLARY_TARGET2 )>0 ) \
		| ((cat['EBOSS_TARGET0'] & bits_EBOSS_TARGET0 )>0 )         \
		| ((cat['EBOSS_TARGET1'] & bits_EBOSS_TARGET1 )>0 )         \
		| ((cat['EBOSS_TARGET2'] & bits_EBOSS_TARGET2 )>0 )
	cat = cat[select]
	print cat.size


	print min(cat[redShiftKey])
	print max(cat[redShiftKey])
	
	## Map
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.plot(cat["RA"], cat["DEC"], linestyle="", marker="o")
	plt.xlabel("Right Ascension (degree)")
	plt.ylabel("Declination (degree)")
	plt.title("BOSS DR12")
	plt.show()
	## Distribution redshift
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.hist(cat[redShiftKey], bins=50)
	plt.xlabel("Z")
	plt.ylabel("#")
	plt.title("BOSS DR12")
	plt.show()

	
	col_ra              = pyfits.Column(name='RA',  format='D', array=cat['RA'], unit='deg')
	col_de              = pyfits.Column(name='DEC', format='D', array=cat['DEC'], unit='deg')
	col_zz              = pyfits.Column(name='Z',   format='D', array=cat[redShiftKey] )
	
	tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
	tbhdu.writeto('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_EBOSS_updated_2016_05_24.fits', clobber=True)
	
	
	return
def getQsoCatalogueAllQSO():

	cat = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR7_DR12_EBOSS_2016_01_08.fits', memmap=True )[1].data

	print cat.size
	cat = cat[ cat['RA']!=0. ]
	print cat.size

	col_ra              = pyfits.Column(name='RA',  format='D', array=cat['RA'], unit='deg')
	col_de              = pyfits.Column(name='DEC', format='D', array=cat['DEC'], unit='deg')
	col_zz              = pyfits.Column(name='Z',   format='D', array=cat['Z'] )

	## Map
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))	
	plt.grid()
	plt.plot(cat["RA"], cat["DEC"], linestyle="", marker="o")
	plt.xlabel("Right Ascension (degree)")
	plt.ylabel("Declination (degree)")
	plt.title("BOSS DR12")
	plt.show()
	## Distribution redshift
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.hist(cat['Z'][ cat['Z']>1.7 ], bins=50)
	plt.xlabel("Z")
	plt.ylabel("#")
	plt.title("BOSS DR12")
	plt.show()
	
	tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
	tbhdu.writeto('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR7_DR12_EBOSS_2016_01_08.fits', clobber=True)
def getQsoCatalogueAllObjects():
	'''

	'''

	path = '/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/'
	listPAth = [path+'QSO_DR7_DR12_EBOSS_2016_01_08.fits',
			path+'DLA_all.fits',
			path+'all_Britt.fits',
			path+'VIPERS.fits',
			path+'QSO_3DHST.fits',
			path+'LOWZ_all.fits',
			path+'CMASS_all.fits']
	name = ['QSO','DLA','Britt','VIPERS','3DHST','LOWZ','CMASS']

	## Distribution redshift
	for i in numpy.arange(len(listPAth)):
		cat = pyfits.open(listPAth[i], memmap=True )[1].data
		cat = cat[ (cat['Z']>0.1) ]
		cat = cat[ (cat['Z']<7.) ]
		if (cat.size==0): continue
		plt.hist(cat['Z'], bins=100,histtype='step',label=name[i])

	plt.xlabel("Z")
	plt.ylabel("#")
	myTools.deal_with_plot(False,False,True)
	plt.show()

	### Merge everyThing
	cat = pyfits.open(listPAth[0], memmap=True )[1].data
	ra = cat['RA']
	de = cat['DEC']
	zz = cat['Z']
	for i in numpy.arange(1,len(listPAth)):
		cat = pyfits.open(listPAth[i], memmap=True )[1].data
		ra = numpy.append(ra, cat['RA'])
		de = numpy.append(de, cat['DEC'])
		zz = numpy.append(zz, cat['Z'])

	## Map
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))	
	plt.grid()
	plt.plot(ra, de, linestyle="", marker="o")
	plt.xlabel("Right Ascension (degree)")
	plt.ylabel("Declination (degree)")
	plt.show()
	## Distribution redshift
	plt.ticklabel_format(style='sci', axis='z', scilimits=(0,0))
	plt.grid()
	plt.hist(zz, bins=200)
	plt.xlabel("Z")
	plt.ylabel("#")
	plt.show()

	### Save	
	col_ra              = pyfits.Column(name='RA',  format='D', array=ra, unit='deg')
	col_de              = pyfits.Column(name='DEC', format='D', array=de, unit='deg')
	col_zz              = pyfits.Column(name='Z',   format='D', array=zz)
	tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
	tbhdu.writeto('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/ALL_EVERY_OBJECTS_2016_01_08.fits', clobber=True)
def create_fits_qso(cat_path, sizeMax):

	### Create a list of QSOs
	col_ra              = pyfits.Column(name='RA',  format='D', array=numpy.zeros(sizeMax), unit='deg')
	col_de              = pyfits.Column(name='DEC',  format='D', array=numpy.zeros(sizeMax), unit='deg')
	col_zz              = pyfits.Column(name='Z',  format='D', array=numpy.zeros(sizeMax), unit='0 (redshift)')

	tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
	tbhdu.writeto(cat_path, clobber=True)

	return
def remove_empty_cell_in_cat(cat_path):

	cat = pyfits.open(cat_path, memmap=True)[1].data
	print '  nb of QSOs before = ', cat.size
	cut = numpy.invert(numpy.logical_and( cat['RA']==0., numpy.logical_and( cat['DEC']==0.,cat['Z'] == 0.)))
	cat = cat[ (cut) ]
	print '  nb of QSOs after  = ', cat.size
	nbQSO = cat.size
	
	pyfits.writeto(cat_path, cat, clobber=True)
	
	return

### Create a catalogue:
cat_path = '/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_DR7_DR12_EBOSS_2016_05_24.fits'
sizeMax  = 2000000
#create_fits_qso(cat_path, sizeMax)
remove_empty_cell_in_cat(cat_path)

#getQsoCatalogueAllObjects()

#getQsoCatalogueEBOSS()
#main()

'''
path = '/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/'
cat = pyfits.open(path + 'QSO_DR7_DR12_EBOSS.fits', memmap=True )[1].data
ra = cat['RA']
de = cat['DEC']
zz = cat['Z']
cat = pyfits.open(path + 'DLA_all.fits', memmap=True )[1].data
ra = numpy.append( ra, cat['RA'])
de = numpy.append( de, cat['DEC'])
zz = numpy.append( zz, cat['Z'])
cat = pyfits.open(path + 'all_Britt.fits', memmap=True )[1].data
ra = numpy.append( ra, cat['RA'])
de = numpy.append( de, cat['DEC'])
zz = numpy.append( zz, cat['Z'])



print ra.size

## Map
plt.plot(ra,  de, linestyle="", marker="o")
plt.xlabel(r'$R.A. (\degree)$')
plt.ylabel(r'$Dec. (\degree)$')
myTools.deal_with_plot(False,False,True)
plt.show()
## Distribution redshift
plt.hist(zz[ numpy.logical_and( zz>1., zz<7. ) ], bins=50)
plt.xlabel(r'$z$')
plt.ylabel(r'$\#$')
myTools.deal_with_plot(False,False,True)
plt.show()


col_ra              = pyfits.Column(name='RA',  format='D', array=ra, unit='deg')
col_de              = pyfits.Column(name='DEC', format='D', array=de, unit='deg')
col_zz              = pyfits.Column(name='Z',   format='D', array=zz )
	
tbhdu = pyfits.BinTableHDU.from_columns([col_ra, col_de, col_zz])
tbhdu.writeto('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/ALL_OBJECTS.fits', clobber=True)
'''

'''
cat3 = pyfits.open('/home/gpfs/manip/mnt/bao/hdumasde/Data/LYA/FitsFile_eBOSS_Guy/all_eBOSS_primery/eBOSS_primery.fits',memmap=True)[1].data
cat3 = cat3[ cat3['Z_VI']>1.7 ]
cat3 = cat3[ cat3['Z_VI']<7. ]
RA  = cat3['RA']
Dec = cat3['DEC']
plt.errorbar(RA,Dec, fmt='o')
plt.show()
'''

"""
import numpy as np
import ephem

org=0
projection='mollweide'

fig = plt.figure(figsize=(10, 5))
ax = fig.add_subplot(111, projection=projection, axisbg ='white')


cat = pyfits.open('/home/gpfs/manip/mnt0607/bao/hdumasde/Data/Catalogue/QSO_ALL_TESTS.fits')[1].data
cat = cat[ cat['Z']>1.7 ]
cat = cat[ cat['Z']<7. ]
RA  = cat['RA']
Dec = cat['DEC']
RA[ RA>180. ] -= 360.
ax.scatter(np.radians(RA),np.radians(Dec), label=r'$DR7+DR12: \, QSO$')  # convert degrees to radians


cat3 = pyfits.open('/home/gpfs/manip/mnt/bao/hdumasde/Data/LYA/FitsFile_DR12_Guy/DR12_primery/DR12_primery.fits',memmap=True)[1].data
cat3 = cat3[ cat3['Z_VI']>2. ]
cat3 = cat3[ cat3['Z_VI']<7. ]
RA  = cat3['RA']
Dec = cat3['DEC']
RA[ RA>180. ] -= 360.
ax.scatter(np.radians(RA),np.radians(Dec), label=r'$DR12: \, Forest$', color='red')  # convert degrees to radians


cat4 = pyfits.open('/home/gpfs/manip/mnt/bao/hdumasde/Data/LYA/FitsFile_eBOSS_Guy/all_eBOSS_primery/eBOSS_primery.fits',memmap=True)[1].data
cat4 = cat4[ cat4['Z_VI']>1.7 ]
cat4 = cat4[ cat4['Z_VI']<7. ]
RA  = cat4['RA']
Dec = cat4['DEC']
RA[ RA>180. ] -= 360.
ax.scatter(np.radians(RA),np.radians(Dec), label=r'$eBOSS \, forest$', color='green')  # convert degrees to radian


lon_array = np.arange(0,360)
lat = 0.
eq_array = np.zeros((360,2))
for lon in lon_array:
    ga = ephem.Galactic(np.radians(lon), np.radians(lat))
    eq = ephem.Equatorial(ga)
    eq_array[lon] = np.degrees(eq.get())

RA  = eq_array[:,0]
Dec = eq_array[:,1]
RA[ RA>180 ] -= 360.
coord = numpy.zeros(shape=(RA.size,2))
sort = RA.argsort()
RA  = RA[sort]
Dec = Dec[sort]
ax.plot(np.radians(RA),np.radians(Dec), color='green', linestyle='dashed', linewidth=5)  # convert degrees to radians

plt.xlabel(r'$R.A. (\degree)$')
plt.ylabel(r'$Dec. (\degree)$')

myTools.deal_with_plot(False,False,True)

plt.legend(bbox_to_anchor=(0., 1., 1., 0.), fontsize=30, frameon=False, scatterpoints=1,ncol=1, loc=4)
plt.show()




"""











